Re-enable tree checkers

VladUreche opened the pull request:
My expectation is that tree checkers are re-typechecking the trees and making sure they are consistent. Unfortunately, following patch aced32d05c97651534f468bc9a475ea5f6ae75b8, the call to clearType() was removed, thus the typer no longer recursed inside the trees, rendering the type checkers framework useless.

This is an attempt to make the tree checkers run again, by resetting the type of a tree before the call to super.typed, thus allowing the typer to actually visit the entire tree (not just the outer package
definition).

The work was prompted by SI-9442, where the type checkers would gladly validate the inconsistent trees.

VladUreche committed:
Re-enable tree checkers

My expectation is that tree checkers are re-typechecking the trees
and making sure they are consistent. Unfortunately, following
patch aced32d05c97651534f468bc9a475ea5f6ae75b8, the call to
clearType() was removed, thus the typer no longer recursed inside
the trees, rendering the type checkers framework useless.

This is an attempt to make the tree checkers run again, by resetting
the type of a tree before the call to super.typed, thus allowing the
typer to actually visit the entire tree (not just the outer package
definition).

The work was prompted by SI-9442, where the type checkers would
gladly allow validate the inconsistent trees.

VladUreche commented:
Ping @scala/team-scala since this is serious: we've been running without consistency checks for 2 years now.

VladUreche commented:
Surprising, all tests passed. Either the ASTs generated by Scala are incredibly consistent or nobody's checking them with `-Ycheck:...`. I suppose it's the 2nd.

Review by @retronym and @SethTisue.

lrytz commented:
@VladUreche i think you put the wrong revision, the one you mentions keeps the call to `clearType` (https://github.com/scala/scala/commit/288b12e48af3222134f3985b65e40f202af81df2#diff-cffe16cefdf13436e44303985f98b38fR267). However, it was actually removed in https://github.com/scala/scala/commit/aced32d05c97651534f468bc9a475ea5f6ae75b8#diff-cffe16cefdf13436e44303985f98b38fL267, which indeed seems accidental.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/TreeCheckers.scala:
```diff
@@ -262,7 +262,14 @@ abstract class TreeCheckers extends Analyzer {
     )
     private def checkedTyped(tree: Tree, mode: Mode, pt: Type): Tree = {
-      val typed = wrap(tree)(super.typed(tree, mode, pt))
+      val typed = wrap(tree)(super.typed(tree.setType(null), mode, pt))
```
Is there a reason to use `setType(null)` instead of `clearType`?

VladUreche commented:
Thanks for the review @lrytz and sorry for including the wrong hash (I didn't notice the `clearType()` on the other side). I just updated the hash and switched to using `clearType()`.

VladUreche commented:
And now it's also rebased on `2.11.x`.

SethTisue commented:
I'm running the community build on this at https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/73/

EDIT: timed out. successor: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/74/consoleFull

VladUreche commented:
Thanks Seth!

SethTisue commented:
green community build: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/74/consoleFull

VladUreche commented:
:+1: thanks @SethTisue

retronym commented:
LGTM

VladUreche commented:
:+1: yaay!
